server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -356,34 +358,46 @@ public TimeValue timeout() {
         });
     }
 
+    private boolean hasPreIndexGensVersionSnapshots(RepositoryData repositoryData, String repositoryName) {
+        final boolean hasOldFormatSnapshots = repositoryData.indexMetaDataGenerations().isEmpty()
+            && hasOldFormatSnapshots(INDEX_GEN_IN_REPO_DATA_VERSION, repositoryName, repositoryData, null);
+        assert hasOldFormatSnapshots == false || repositoryData.indexMetaDataGenerations().isEmpty() :
+            "Found non-empty index generations [" + repositoryData.indexMetaDataGenerations()
+                + "] but repository contained old version snapshots";
+        return hasOldFormatSnapshots;
+    }
+
     public boolean hasOldVersionSnapshots(String repositoryName, RepositoryData repositoryData, @Nullable SnapshotId excluded) {
-        final Collection<SnapshotId> snapshotIds = repositoryData.getSnapshotIds();
-        final boolean hasOldFormatSnapshots;
-        if (snapshotIds.isEmpty()) {
-            hasOldFormatSnapshots = false;
-        } else {
-            if (repositoryData.shardGenerations().totalShards() > 0) {
-                hasOldFormatSnapshots = false;
-            } else {
-                try {
-                    final Repository repository = repositoriesService.repository(repositoryName);
-                    hasOldFormatSnapshots = snapshotIds.stream().filter(snapshotId -> snapshotId.equals(excluded) == false).anyMatch(
-                        snapshotId -> {
-                            final Version known = repositoryData.getVersion(snapshotId);
-                            return (known == null ? repository.getSnapshotInfo(snapshotId).version() : known)
-                                .before(SHARD_GEN_IN_REPO_DATA_VERSION);
-                        });
-                } catch (SnapshotMissingException e) {
-                    logger.warn("Failed to load snapshot metadata, assuming repository is in old format", e);
-                    return true;
-                }
-            }
-        }
+        final boolean hasOldFormatSnapshots = repositoryData.shardGenerations().totalShards() == 0
+            && hasOldFormatSnapshots(SHARD_GEN_IN_REPO_DATA_VERSION, repositoryName, repositoryData, excluded);
         assert hasOldFormatSnapshots == false || repositoryData.shardGenerations().totalShards() == 0 :
             "Found non-empty shard generations [" + repositoryData.shardGenerations() + "] but repository contained old version snapshots";
         return hasOldFormatSnapshots;
     }
 
+    private boolean hasOldFormatSnapshots(Version minVersion, String repositoryName, RepositoryData repositoryData,
```

Comment:
I realise that this will iterate the snapshot list twice worst case but note that it will never actually make any API requests in case there is snapshots that all support shard generations but not index meta generations because all versions that have shard generations also write the `version` field to `RepositoryData`.
I just figured I'd use the same code for iterating for both versions because this fact would be somewhat confusing if used outright (IMO).